chain-extensions/unified-accounts/src/lib.rs

```diff
@@ -0,0 +1,84 @@
+// This file is part of Astar.
+
+// Copyright (C) 2019-2023 Stake Technologies Pte.Ltd.
+// SPDX-License-Identifier: GPL-3.0-or-later
+
+// Astar is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Astar is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Astar. If not, see <http://www.gnu.org/licenses/>.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use astar_primitives::evm::{EvmAddress, UnifiedAddressMapper};
+use core::marker::PhantomData;
+use sp_runtime::DispatchError;
+
+use frame_support::{traits::Get, DefaultNoBound};
+use pallet_contracts::chain_extension::{
+    ChainExtension, Environment, Ext, InitState, Result as DispatchResult, RetVal,
+};
+use parity_scale_codec::Encode;
+pub use unified_accounts_chain_extension_types::Command::{self, *};
+
+#[derive(DefaultNoBound)]
+pub struct UnifiedAccountsExtension<T, UA>(PhantomData<(T, UA)>);
+
+impl<T, UA> ChainExtension<T> for UnifiedAccountsExtension<T, UA>
+where
+    T: pallet_contracts::Config + pallet_unified_accounts::Config,
+    UA: UnifiedAddressMapper<T::AccountId>,
+{
+    fn call<E>(&mut self, env: Environment<E, InitState>) -> DispatchResult<RetVal>
+    where
+        E: Ext<T = T>,
+    {
+        let mut env = env.buf_in_buf_out();
+        match env.func_id().try_into().map_err(|_| {
+            DispatchError::Other("Unsupported func id in Unified Accounts Chain Extension")
+        })? {
+            GetEvmAddress => {
+                let account_id: T::AccountId = env.read_as()?;
+
+                let base_weight = <T as frame_system::Config>::DbWeight::get().reads(1);
+                env.charge_weight(base_weight)?;
+                // write to buffer
+                UA::to_h160(&account_id).using_encoded(|r| env.write(r, false, None))?;
+            }
+            GetEvmAddressOrDefault => {
+                let account_id: T::AccountId = env.read_as()?;
+
+                let base_weight = <T as frame_system::Config>::DbWeight::get().reads(1);
+                env.charge_weight(base_weight)?;
+                // write to buffer
+                UA::to_h160_or_default(&account_id).using_encoded(|r| env.write(r, false, None))?;
```

Comment:
It's probably a minor concern, but shouldn't the return value also carry the information if this is mapped or just default value?

Reply:
User (contract) can know whether the address is mapped or not using the current interface, like this
```rust
// know whether the address is mapped or not
let is_mapped = UAExtension::to_h160().is_some();
// get the address
let evm_address = UAExtension::to_h160_or_default();
```

Reply:
Yeah it'll take 2 db reads.  

Personally, I just want it to closely follow the interface of existing `UnifiedAddressMapper` trait for both CE and Precompile, it's easy to maintain that way. So if we add /change interface for this, it should reflect into trait as well.

We can change it to like this? happy to hear if someone has other ideas for interface.
```rust
let (address, is_mapped) = UAExtension::to_h160_or_default();
```

Reply:
> Personally, I just want it to closely follow the interface of existing UnifiedAddressMapper trait for both CE and Precompile, it's easy to maintain that way. 

Actually, I disagree with this. I think we should do quite the opposite - changing the precompile/chain-extension should be avoided at all cost. For things like UA, the existing interface function should never change or stop working.

Separating the runtime implementation from CE/precompiles is better therefore. E.g. for dApp staking precompile, we explicitly avoided returning types that were/are used in the pallet. This way maintenance is actually much easier since if you refactor the pallet, you don't need to modify the interface, only the implementation.

Reply:
I see, but I meant in backward manner, like if this is needed (efficiently) by smart contracts then it's a good idea to add the same functionality to UA trait.

For CE interface, any suggestions? (cc: @PierreOssun)

Reply:
I'd only suggest to modify the return type for now.
It's useful information if this is default value, or _really_ mapped one.
Not sure what others think about this, but it can also be made into an issue.

The comment about interface was more general.

Reply:
It makes sense to me to add `is_mapped` to the return value. With full info provided it's unlikely we need to update the interface in the future.